CloneTest/Program.cs

```diff
@@ -1,5 +1,214 @@
+            //double[] density_distribution = new double[no_of_hubs];
+            //double[] desirability_distribution = new double[no_of_hubs];
+
+            Random rnd = new Random();
+            for (int i = 0; i < no_of_hubs; ++i)
+            {
+                var x = rnd.Next(0, density_size_x);
+                var y = rnd.Next(0, density_size_y);
+                hubs.Add(Tuple.Create(x, y));
+                hubs_desity.Add(Tuple.Create(x, y, density_map[x, y]));
+                hubs_desirability.Add(Tuple.Create(x, y, desirability_map[x, y]));
+
+            }
+
+            var density_distribution = PrepareHubDistribution(hubs_desity);
+            var desirability_distribution = PrepareHubDistribution(hubs_desirability);
+
+            //Console.WriteLine("Density map");
+            //Print2DMap(density_map, density_size_x, density_size_y);
+
+            //Console.WriteLine("Desirability map");
+            //Print2DMap(desirability_map, density_size_x, density_size_y);
+
+            var density_dist_max = density_distribution.Max();
+            var desirability_dist_max = desirability_distribution.Max();
+
+            var density_points = new List<int>();
+            var desirability_points = new List<int>();
+
+            for (int point_no = 0; point_no < no_of_links; ++point_no)
+            {
+                var density_val = rnd.Next(0, (int)density_dist_max + 1);
+                var desirability_val = rnd.Next(0, (int)desirability_dist_max + 1);
+
+                var density_id = BinarySearch(density_distribution, density_val);
+                var desirability_id = BinarySearch(desirability_distribution, desirability_val);
+
+                if (hubs_desity[density_id].Item1 == hubs_desirability[desirability_id].Item1 &&
+                    hubs_desity[density_id].Item2 == hubs_desirability[desirability_id].Item2)
+                {
+                    continue;
+                }
+                density_points.Add(density_id);
+                desirability_points.Add(desirability_id);
+            }
+
+            for (int i = 0; i < density_points.Count(); ++i)
+            {
+                int density_id = density_points[i];
+                int desirability_id = desirability_points[i];
+                Console.WriteLine("Travel from x: " + hubs_desity[density_id].Item1 +
+                    " y: " + hubs_desity[density_id].Item2 +
+                    " density: " + hubs_desity[density_id].Item3 +
+                    ", to x: " + hubs_desirability[desirability_id].Item1 +
+                    " y: " + hubs_desirability[desirability_id].Item2 +
+                    " desirability: " + hubs_desirability[desirability_id].Item3);
+            }
+
+        }
+
+        static int ReadInt()
+        {
+            string? linia = Console.ReadLine();
+            if (linia == null)
+            {
+                throw new ArgumentException("Oczekiwano wartosci typu int");
+            }
```

Comment:
Podejście dobre, ale to jednak 4 linijki kodu. W tej sytuacji polecam ci użyć 
```csharp
linia ?? throw new ArgumentException("Oczekiwano wartosci typu int");
```
oszczędzasz 3 linijki kodu, a operator `??` działa na zasadzie "jeśli wartość z lewej == null to wykonaj kod z prawej strony. Inną sprawą jest fakt, że jeszcze nigdy nie udało mi się uzyskać z konsoli null-a. W przypadku wciśnięcia entera dostajesz pusty string

Reply:
Gdy użyłem linii ```linia ?? throw new ArgumentException("Oczekiwano wartosci typu int");``` otrzymałem błąd: "CS0201". Zrobiłem ```_ = linia ?? throw new ArgumentException("Oczekiwano wartosci typu int");```.

Nie byłem pewny jak to działa w C#. Co by się stało gdyby na przykład przekierować wyjście jednego programu do drugiego przez potok.